### ISSUE TITLE
Patternmatcher's unexpected behaviors with logical links

### ISSUE THREAD

**Issue by Eman22S** [opencog/atomspace#2582]
There are few problems I have came across while using bindlink to instantiate a hypergraph of Orderedlink. 
1. It seems to get into an infinite loop when I have this rule 
```
(define and-identity
   (BindLink
     (Variable "$op")
     (AndLink
       (Variable "$op")
       (True))

     (Ordered
      (AndLink
         (Variable "$op")
         (False)))
 ))
```
run in an atomspace that contains 
```
(AndLink
  (False)
  (True))
```
using `cog-execute! `

2. It does not subsitute glob when we have the above rule with the globnode instead of variable node
```
(define and-identity
   (BindLink
     (Glob "$op")
     (AndLink
       (Glob "$op")
       (True))

     (Ordered
      (AndLink
         (Glob "$op")
         (False)))
 ))
 ```

Rather it produces this as an output
```
$2 = (SetLink
   (OrderedLink
      (AndLink
         (GlobNode "$op")
         (FalseLink
         )
      )
   )
)
```

**Comment by linas** [comment]
>  infinite loop 

Sure. Sounds plausible. It's not hard to write infinite loops. And it seems that's what you did, so not a surprise.  It is impossible to detect infinite loops -- this is known to be a problem not solvable by Turning machines. The pattern matcher can detect and halt a few of simpler the simpler inf loops, but can't do so in general.  Apparently, it fails to detect the one you wrote.

To avoid infinite loops, you should type your variables, and I am guessing that you meant to use  `PresentLink` 00 which is what you need to find something to be contained in the atomspace. Withtout the `PresentLink`, the TrueLink is merely true. rather than being searched for. That is,  your first query is equivalent to 
```
(define and-identity
   (BindLink
     (Variable "$op")
     (AndLink
       (Variable "$op"))
     (Ordered
      (AndLink
         (Variable "$op")
         (False)))
 ))
```
which is equivalent to
```
(define and-identity
   (BindLink
     (Variable "$op")
     (Present (Variable "$op"))
     (Ordered
      (AndLink
         (Variable "$op")
         (False)))
 ))
```
which is no obviously an inf loop.

> (Glob

That's odd. It should have been an inf loop also.  I'm not sure what it wasn't. I'll look into that &fix it..

Inf loops can be useful, when they are tail-recursive -- you can write a variety of different atomese inf loops, that are useful for handling events, atomspace updates, reactions to incoming stimulus, etc.

**Comment by linas** [comment]
fixed in #2584 -- both cases are now inf loop.

Look in your log files: you should see this warning printed:
https://github.com/opencog/atomspace/blob/1c21d91fee67d65078da1340274d45fc816e68e8/opencog/query/InitiateSearchCB.cc#L728-L751

**Comment by linas** [comment]
Closing; first part of the bug report is a user error. Second part should have resulted in inf recursion, and now it does, fixed in #2584

**Comment by linas** [comment]
p.s. Sorry for the abrupt reply. You are maybe the 6th or 8th person who has re-discovered these inf loops and was surprised by them.  There has been a lot of confusion about how they happen. (inf loops are fixed points of recursive functions) In general, typing is a good thing, and failing to specify types for search patterns is ... bad. In several ways... without a type, the pattern matcher has to assume a worst-case scenario and search for everything. 

But there was another hole you fell into: the pattern matcher is actually a satisfiability solver -- it is trying to satisfy all of the clauses in the pattern.  These clauses can evaluate to "true" or "false", and can be and/or/not'ed together.  A very special clause is `PresentLink` which says "is this atom present in the atomspace?"  That is the one that you want to use, as otherwise TrueLink/FalseLink/AndLink/OrLink/NotLink are just evaluated, as they stand. You don't see `PresentLink` a lot in the examples, because it is implicit in most cases: if someone says `(Concept "foo")`, you can be sure they meant (Present (Concept "foo"))` so that's automatic. But for True/False/etc it cannot be automatic, and that is what you tripped over...
